manager/allocator/network/ipam/ipam.go

```diff
@@ -0,0 +1,831 @@
+			// is nothing to do
+			len(nw.IPAM.Configs) == 0 {
+			continue
+		}
+		local := &network{
+			nw:        nw,
+			pools:     make(map[string]string),
+			endpoints: make(map[string]string),
+		}
+		// TODO(dperny): is it possible for a network before this change to
+		// have an empty ipamName but be allocated?
+		ipamName := nw.IPAM.Driver.Name
+		ipamOpts := nw.IPAM.Driver.Options
+		if ipamOpts == nil {
+			ipamOpts = map[string]string{}
+		}
+		// if we have an ipam driver name and we have IPAM configs, but for
+		// some reason we don't have IPAM options, just fill that field in.
+		// shouldn't do any harm. i doubt we'll ever hit this in prod
+		local.nw.IPAM.Driver.Options = ipamOpts
+		// IPAM returns the Ipam object and also its capabilities. We don't
+		// use the capabilities or care about them right now, so just ignore
+		// that part of the return value
+		ipam, _ := a.drvRegistry.IPAM(ipamName)
+		if ipam == nil {
+			return errors.ErrBadState("ipam driver %q cannot be found", ipamName)
+		}
+		_, addressSpace, err := ipam.GetDefaultAddressSpaces()
+		// the only errors here result from having an invalid ipam driver
+		if err != nil {
+			return errors.ErrInternal("ipam %v for network %v returned an error when requesting the default address space: %v", ipamName, nw.ID, err)
+		}
+
+		// now initialize the IPAM pools. IPAM pools are the set of addresses
+		// available for a specific network. There is one IPAM config for every
+		// pool. In order for this restore operation to be consistent, a
+		// network must have no ipam configs if it hasn't been allocated
+		for _, config := range local.nw.IPAM.Configs {
+			// the last param of RequestPool is "v6", meaning IPv6, which we
+			// don't support, hence passing "false"
+			poolID, poolIP, _, err := ipam.RequestPool(addressSpace, config.Subnet, config.Range, ipamOpts, false)
+			if err != nil {
+				// typically, if there was an error in requesting a pool, we
+				// would release the pools we've already allocated. However, if
+				// there is an error at this stage, it means the whole object
+				// store is in a bad state, so we don't do that, we just
+				// abandon everything.
+				return errors.ErrBadState(
+					"error reserving ipam pool for network %v: %v",
+					nw.ID, err,
+				)
+			}
+			// each IPAM pool we use has a separate ID referring to it. We also
+			// keep a map of each IP address we have allocated for this network
+			// and what pool it belongs to, so we can deallocate an address
+			// from the correct pool
+			local.pools[poolIP.String()] = poolID
+			// now we just need to reserve the gateway address for this pool.
+			// set the IPAM request address type to "Gateway" to tell the IPAM
+			// driver that's the kind of address we're requesting
+			ipamOpts[ipamapi.RequestAddressType] = netlabel.Gateway
+			// and now, if we have a Gateway address for this network, we need
+			// to allocate it. This condition should never happen; a network
```

Comment:
Should this say "this condition should always be true", the condition being "we have a gateway address for this network"?